### ISSUE TITLE
Heavy compositing uses the wrong cached frame during render

### ISSUE THREAD

**Issue by rainChu** [rlguy/Blender-FLIP-Fluids#489]
#### System Information

**Blender Version (including hash):** 2.81a (f1aa4d18d49d)
**FLIP Fluids Version:** 1.0.7
**Operating System:** Windows 10
**CPU:** Intel i7-4790 @ 3.60GHz
**GFX:** GTX 1060, 6GB
**RAM:** 16GB

#### Describe the bug

When the compositor is used heavily, the fluid cache might be offset by -1 during render, causing all fluid to display one frame earlier than everything else in the scene. This doesn't affect realtime playback.

#### To Reproduce

Download the following .blend file which contains a minimum setup: [simplified_heavy_compositing.blend](https://rainChu.com/s/simplified_heavy_compositing.blend)

1. Bake the fluid cache.
2. Scrub through the animation in the viewport. Pay particular attention to the frame the fluid enters the world at, frame 12.
3. Be sure "Composited Cycles" Scene is selected. This scene uses the compositor in a way which causes the bug.
4. Render frame 11. The fluid from frame 12 should be visible.
5. Render an image from "Simple Cycles" and "Composited Eevee" scenes. These work as expected, but are otherwise identical (all aspects are fully linked. The only difference is compositor is turned off.)

**Comment by rainChu** [comment]
I just want to append in the comments that this isn't caused by the patch you just gave; I haven't patched my files yet. This occurs in the stable addon from Blender Market

**Comment by rlguy** [comment]
Hi rainChu,

Thanks for the report and .blend file! I was able to reproduce this issue with the instructions.

This is an interesting bug. It looks like this is caused by Blender behavior that I did not expect. I don't have in-depth knowledge of how the compositor works, but it seems that Blender is triggering multiple frame changes during render when the compositor is used.

Here are the frame changes using **Composited Cycles**:

```
Begin frame render
1st Frame change: 11
2nd Frame change: 11
3rd Frame change: 11
4th Frame change: 11
5th Frame change: 11
6th Frame change: 10
7th Frame change: 12
8th Frame change: 11
End frame render
```

So the problem is that Blender is triggering a change to frame 12 and that is the frame that is being used when rendering the fluid. I am not sure what is causing the change to frames 10 and 12, but perhaps the compositor needs info from the previous and next frame.

Here are the frame changes using **Composited Eevee**:

```
Begin frame render
1st Frame change: 12
2nd Frame change: 12
3rd Frame change: 12
4th Frame change: 12
5th Frame change: 12
End frame render
```

All of the frame changes are consistent, so maybe Eeevee does not need data from the previous/next frames.

And here are the frame changes using **Simple Cycles**:

```
Begin frame render
1st Frame change: 11
End frame render
```

Just a single simple frame change.

## Bug fix

I have fixed this issue for the next version. At the beginning of a frame render, the addon will lock the  fluid simulation frame so that it cannot be changed in the middle of a frame render. This will make the loaded fluid frame consistent throughout the render stages during the frame. This fix will only lock the frame loaded by the simulator in order to not affect other areas of Blender that require a frame change to the previous/next frame.

## Workaround

This fix is a bit more complicated than issue #488, so I am unable to provide a single file patch to fix this in your installation. A workaround for now would be to use the patch from #488 and offset the frame.

\- Ryan
